### PR TITLE
View attachments

### PR DESCRIPTION
**Pull request by alloy**
Still rough around the edges, but good enough for the first beta.

view-attachment

Skip New Tests.

### CODE REVIEW

File: src/lib/Components/Inbox/Conversations/Message.tsx [hunk 1]
```diff
@@ -119,6 +131,9 @@ export default Relay.createContainer(Message, {
           content_type
           download_url
           file_name
+          id
+          ${ImagePreview.getFragment("imageAttachment")}
+          ${PDFPreview.getFragment("pdfAttachment")}
```
Comment (alloy) [inline src/lib/Components/Inbox/Conversations/Message.tsx]: You weren’t letting Relay handle the fetching of the data for these component, but were fetching it yourself and then pushing that into components that Relay expected to manage, hence the warning.

**Comment by ArtsyOpenSource** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  No CHANGELOG added.
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  Missing Test Files:

- `src/lib/NativeModules/__tests__/SwitchBoard-tests.tsx`

If these files are supposed to not exist, please update your PR body to include "Skip New Tests".
  </td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: Example/Podfile.lock [hunk 2]
```diff
@@ -173,13 +173,13 @@ EXTERNAL SOURCES:
     :branch: fetch-user-details
     :git: https://github.com/artsy/Artsy-Authentication.git
   Emission:
-    :path: ../
+    :path: "../"
```
Comment (mzikherman) [inline Example/Podfile.lock]: The battle continues!

File: src/lib/Components/Inbox/Conversations/Message.tsx [hunk 3]
```diff
@@ -58,19 +60,29 @@ interface Props extends RelayProps {
...
 export class Message extends React.Component<Props, any> {
-  renderAttachmentPreviews(attachments) {
+  renderAttachmentPreviews(attachments: Props["message"]["attachments"]) {
+    const previewAttachment = attachmentID => {
+      const attachment = attachments.find(({ id }) => id === attachmentID)
+      SwitchBoard.presentMediaPreviewController(
+        this,
+        attachment.download_url,
+        attachment.id,
+        attachment.file_name.split(".").reverse()[0]
```
Comment (mzikherman) [inline src/lib/Components/Inbox/Conversations/Message.tsx]: 😎

File: Example/Podfile.lock (hunk 2 above)
Comment (alloy) [inline Example/Podfile.lock]: Yeah… I’m on Ruby 2.3.0 in this project. I guess one thing we could do is consolidate on a Ruby version (and add a `.ruby-version` file) ?

File: src/lib/Components/Inbox/Conversations/Message.tsx [hunk 4]
```diff
@@ -58,19 +60,29 @@ interface Props extends RelayProps {
...
+      )
+    }
+
     return attachments.map(attachment => {
       if (attachment.content_type.startsWith("image")) {
         return (
-          <ImagePreviewContainer>
-            <ImagePreview imageAttachment={attachment} />
+          <ImagePreviewContainer key={attachment.id}>
```
Comment (alloy) [inline src/lib/Components/Inbox/Conversations/Message.tsx]: @mzikherman This was another of the warnings you were getting. When adding a list of child components you need to specify the `key` prop with a unique value for each child, preferably a stable value (not an index). See https://facebook.github.io/react/docs/lists-and-keys.html

**Comment by alloy** [comment]
It’s especially lacking a download progress indicator for when an attachment has not yet been cached. I was thinking of a circular masked progress indicator (which I _think_ WhatsApp uses?) that would overlay the attachment component, something like [this](http://cl.ly/DxLM) except not as a window-wide modal, however that probably wouldn’t work well for PDF components (unless we’d preview those too).

@katarinabatina can you give some thought into what this should look like?

File: src/lib/Components/Inbox/Conversations/Message.tsx (hunk 3 above)
Comment (alloy) [inline src/lib/Components/Inbox/Conversations/Message.tsx]: No longer doing this :)

File: src/lib/Components/Inbox/Conversations/Message.tsx (hunk 4 above)
Comment (mzikherman) [inline src/lib/Components/Inbox/Conversations/Message.tsx]: 👍

**Comment by mzikherman** [comment]
I love the attachment component refactor.

I'm going to merge in order to rebase some existing PRs, but def would appreciate another set of 👁 on the native code!

File: src/lib/Components/Inbox/Conversations/ConversationSnippet.tsx [hunk 5]
```diff
@@ -107,6 +107,12 @@ export class ConversationSnippet extends React.Component<Props, any> {
     const conversation = this.props.conversation
     const artwork = conversation.artworks[0]
 
+    // TODO We need to make artworks available even if they are unpublished.
+    if (!artwork) {
+      console.warn(`Unable to load artwork for conversation with ID ${conversation.id}`)
+      return null
+    }
```
Comment (alloy) [inline src/lib/Components/Inbox/Conversations/ConversationSnippet.tsx]: This is the part that needs to be taken out once we serve unpublished artworks.

**Pull request merged**